libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -940,13 +962,29 @@ void IRGenerator::generateConstructors(ContractDefinition const& _contract)
 std::string IRGenerator::deployCode(ContractDefinition const& _contract)
 {
 	Whiskers t(R"X(
-		let <codeOffset> := <allocateUnbounded>()
-		codecopy(<codeOffset>, dataoffset("<object>"), datasize("<object>"))
-		<#immutables>
-			setimmutable(<codeOffset>, "<immutableName>", <value>)
-		</immutables>
-		return(<codeOffset>, datasize("<object>"))
+		<?eof>
+			let dataSectionMemoryStart = 0
+			let dataSectionMemorySize = 0
+			<?library>
+				dataSectionMemoryStart := <allocateUnbounded>()
+				dataSectionMemorySize := 32
```

Comment:
You set values in both branches, so you don't have to predeclare these variables.

Also, `auxData` is shorter and IMO also clearer than `dataSectionMemory`.

```suggestion
			<?library>
				let auxDataStart := <allocateUnbounded>()
				let auxDataSize := 32
```

Reply:
Similarly, I'd rename `immutableDataSectionOffset` to `immutableOffset` above. The offset is the same in memory and in auxdata anyway. And currently it sounds a bit like the offset of the start of the data section meant for immutables.

Reply:
You could also simplify this by adding a helper like `pushLibraryAddressImmutable()` to the context (with an assert that it's the first immutable). You wouldn't have to treat libraries specially here and it would even be a tiny bit more efficient.

Reply:
Do you mean to add library address to `reservedMemory`?

Reply:
Yeah, though I now realized than when I wrote it I was thinking about a function that pushes the value directly there. That won't work of course, because the value is only known at runtime (via `address()`). So it would instead be a function that registers the immutable (e.g. `registerLibraryAddressImmutable()`, without args).

And you won't be able to put it in the usual map from variables to offsets with other immutables, so you need a special flag in the context to indicate that this spot is reserved. You could always make it `0`, or, for more generality, actually store an offset and return it with something like `libraryAddressImmutableOffset()`. But in the end it will always end up being `0` anyway because the implementation will only register that single immutable for libraries.

Reply:
OK. But the value of the address in this spot needs to be initialized. This can be only done in the template using `address()` so we still need something like:
```
<?library>
  mstore(<libaray_address_offset_in_memory>, address())
<!library>
```
right?

Reply:
Yes. But after that you can use both uniformly.

This also needs to happen before the reserved memory gets reserved (so before `memoryguard` is inserted), but that's not the problem with normal immutables, so shouldn't be a problem here either. Something to look at when I'll be doing the next review pass.

Reply:
Yes. Just wanted to confirm.